rfcs/053-lambda-logging/README.md

```diff
@@ -0,0 +1,63 @@
+- Lambda invocation (duration assumed trivial, $0.20/million log lines)
+- Kinesis (optional): ($0.04/hr + \$0.12/GB)
+- Network egress (approximately constant)
+
+![architecture for streaming from cloudwatch](https://user-images.githubusercontent.com/4429247/203840497-cb313e4e-7fb1-4b0e-bf01-e8c81dd6c432.png)
+
+**2. Lambda extensions**
+
+[Lambda extensions](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-extensions-api.html) allow code to be integrated into the Lambda execution environment, hooking into the Lambda lifecycle and allowing access to logs and other telemetry. Streaming logs elsewhere is exactly their intended use case - there are already [3rd party extensions](https://docs.datadoghq.com/serverless/libraries_integrations/extension/) for doing this (unfortunately, not for Elasticsearch).
+
+If we built our own Lambda extension, we would expose a Terraform module for provisioning Lambdas that included the extension and necessary config, IAM roles etc etc. We could write it in a language of our choice and use it with any native (ie non-Docker lambda). 
+
+It could potentially use fluent bit for streaming the logs, instead of implementing that manually, although it looks like [others](https://github.com/aws-samples/aws-lambda-extensions/pull/42) have struggled with losing logs for short-lived invocations (probably due to buffering, and possibly resolveable with use of Lambda lifecycle hooks).
+
+**There is a caveat with this option** - it doesn't work transparently for containerised (ie non-native) Lambdas. Instead, the packaged extension has to be [added to the container image](https://docs.aws.amazon.com/lambda/latest/dg/extensions-configuration.html). 
+
+We use containerised Lambdas in the concepts pipeline: we would either have to switch to using native Java  runtimes, or provide our own base images which include the extension code.
+
+There would be no additional cost associated with using a Lambda extension, other than the necessary network egress.
+
+![Native lambda extensions architecture](https://user-images.githubusercontent.com/4429247/203977301-cb6c2cdb-685e-4b62-9ef1-5396fc349b12.png)
+
+## A pragmatic path forward
+
+*Option (1) does not preclude option (2) as further work*: we would already be providing a Terraform module for consumers, and if the Elastic Serverless Forwarder proved insufficient, we would already have written a module to stream the correct format of logs to Elasticsearch. 
+
+This fact clarifies that option (1) - streaming of logs from CloudWatch - is the better starting point. If it proves insufficient or impractical, the work we will have done already will mostly still be useful for option (2).
+
+The work plan might look something like this:
+
+- Provision an Elastic serverless forwarder
+- Create a log group subscription and verify that the forwarder works
+- Create a Terraform module for Lambdas with log groups that subscribe to the forwarder
+- Roll this out across our Lambdas
```

Comment:
Is this a single forwarder for all our accounts, or would we have a serverless forwarder per-account?

I'm envisioning something like https://github.com/wellcomecollection/platform-infrastructure/blob/main/monitoring/terraform/slack_alert_on_lambda_errors.tf but for forwarders.

Reply:
That is sort of to-be-determined by whether we'd use Kinesis. I'll make that explicit.